Sources/CacheAdvance/CacheAdvance.swift

```diff
@@ -0,0 +1,194 @@
+//
+//  Created by Dan Federman on 11/9/19.
+//  Copyright © 2019 Dan Federman.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+/// A performant on-disk cache that can be appended to one element at a time.
+public final class CacheAdvance<T: Codable> {
+
+    // MARK: Initialization
+
+    /// Creates a CacheAdvance.
+    /// - Parameters:
+    ///   - file: The URL for the on-disk cache to live.
+    ///   - maximumBytes: The maximum size of the cache, in bytes.
+    ///   - shouldRoll: When `true`, new logs will overwrite the oldest logs when the cache runs out of space.
+    public init(
+        file: URL,
+        maximumBytes: UInt32,
+        shouldRoll: Bool)
+        throws
+    {
+        self.file = file
+        self.maximumBytes = maximumBytes
+        self.shouldRoll = shouldRoll
+
+        writer = try FileHandle(forWritingTo: file)
+        reader = try FileHandle(forReadingFrom: file)
+
+        lengthOfMessageSuffix = {
+            if shouldRoll {
+                /// We store both the `endOfNewestMessageMarker` and `offsetInFileOfOldestMessage` after each message.
```

Comment:
I personally avoid `we` and use write in the other tense whose name escapes me right now haha. "The message suffix requires space for both `endOfNewestMessageMarker` and `offsetInFileOfOldestMessage` after each message when rolling is enabled"